### ISSUE TITLE
"Should -HaveParameter 'param' -Not -Mandatory" passes tests if parameter is not present

### ISSUE THREAD

**Issue by enoorden** [pester/Pester#2105]
## General summary of the issue
"Should -HaveParameter 'param' -Not -Mandatory" passes tests if parameter is not present

Would expect it to fail

## Describe your environment
Pester version     : 5.3.1 C:\Users\___\Documents\PowerShell\Modules\pester\5.3.1\Pester.psm1
PowerShell version : 7.2.0
OS version         : Microsoft Windows NT 10.0.19043.0

## Steps to reproduce
```powershell
function hi {
    param(
        [parameter(mandatory)]
        [string]$name
    )
    "hi $name"
}

Context 'ctx' {
    It 'test missing param' {
        $fnc = Get-Command 'hi'
        $fnc | Should -HaveParameter 'anotherParam' -Not -Mandatory
    }
}
```

```
Context ctx
  [+] test missing param 6ms (3ms|3ms)
```
-->

## Expected Behavior
would expect test to fail

## Current Behavior
test succeeds

## Possible Solution? (optional)
use two tests
```powershell
Context 'ctx' {
    It 'test missing param' {
        $fnc = Get-Command 'hi'
        $fnc | Should -HaveParameter 'anotherParam'
        $fnc | Should -HaveParameter 'anotherParam' -Not -Mandatory
    }
}
```

**Comment by nohwnd** [comment]
Thanks for reporting this, @lipkau wanna fix this?

**Comment by johlju** [comment]
> We're unable block -NotMandatory -Mandatory now due to how Should works

Would it work saying `Should -HaveParameter -Mandatory:$false` 🤔

**Comment by fflaten** [comment]
> Would it work saying `Should -HaveParameter -Mandatory:$false` 🤔

Hmm, we do have a `$BoundParameters`-dictionary in the function so I guess we could separate undefined (switch defaults to false) from `-Mandatory:$false` without any breaking changes

**Comment by krokofant** [comment]
> `Should -HaveParameter -Mandatory:$false` 
Supporting this would enable the following structure for -ForEach tests:
```powershell
    It 'Has parameter <_.Name>' -TestCases @(
        @{ Name = 'Foo'; Mandatory = $true }
        @{ Name = 'Bar'; }
    ) {
        $command | Should -HaveParameter $Name -Mandatory:([bool]$Mandatory) -Type $Type
    }
```
Where I want to make sure `Foo` is mandatory but `Bar` is not.

**Comment by krokofant** [comment]
There are a lot of nots and not negate in this code so it's a brain melter.

I think instead of this (current):
```powershell
if ($Mandatory) {
    $testMandatory = $attributes | & Where-Object { $_ -is [System.Management.Automation.ParameterAttribute] -and $_.Mandatory }
    $filters += "which is$(if ($Negate) {" not"}) mandatory"

    if (-not $Negate -and -not $testMandatory) {
        $buts += "it wasn't mandatory"
    }
    elseif ($Negate -and $testMandatory) {
        $buts += "it was mandatory"
    }
}
```
It should be this:
```powershell
if ($PSBoundParameters.Keys -contains "Mandatory") {
    $testMandatory = $attributes | & Where-Object { $_ -is [System.Management.Automation.ParameterAttribute] -and $_.Mandatory }
    $filters += "which is$(if ($Negate -or -not $Mandatory) {" not"}) mandatory"

    if (-not ($Negate -or -not $Mandatory) -and -not $testMandatory) {
        $buts += "it wasn't mandatory"
    }
    elseif (($Negate -or -not $Mandatory) -and $testMandatory) {
        $buts += "it was mandatory"
    }
}
```

Can it be simplified a bit? 🤔

**Comment by fflaten** [comment]
> > `Should -HaveParameter -Mandatory:$false`
> 
> Supporting this would enable the following structure for -ForEach tests:
> 
> ```powershell
>     It 'Has parameter <_.Name>' -TestCases @(
>         @{ Name = 'Foo'; Mandatory = $true }
>         @{ Name = 'Bar'; }
>     ) {
>         $command | Should -HaveParameter $Name -Mandatory:([bool]$Mandatory) -Type $Type
>     }
> ```
> 
> Where I want to make sure `Foo` is mandatory but `Bar` is not.

That already works. The inconvenience is only when `-Not` is used at the same time.

Both samples in comment above are breaking changes when parameter is mandatory but you run `Should -HaveParameter abc` (not defined `-Mandatory`). We need to ensure we have test-coverage first before any changes. The `$PSBoundParameters.ContainsKey()` should probably be in the inner if-else.

_**Note to future assignee:**
Any change before new Should/Assert are introduced should also be applied to `$HasArgumentCompleter` + ensure we have tests for all scenarios regarding (with/without `-Not`, parameter exists vs not etc.) to avoid regression later._

**Comment by krokofant** [comment]
> That already works, but I'd recommend being explicit with `Mandatory = $false` in the second testcase. The inconvenience is only when `-Not` is used at the same time.

The second case does not work from what I've tried without the changes I mentioned. Since the current Pester code just has an if  that checks if `$Mandatory` is true. The explicit `-Mandatory:$false` is never tested.

I thank you for the recommendation, but please re-iterate what you mean should work. `-Mandatory:([bool]$false)` does not seem to work in the sense that it should break if the parameter is not mandatory. It just skips checking anything if $Mandatory is false.

**Comment by fflaten** [comment]
Yeah, you're absolutely right. Looked at the wrong sample when I answered.
